utils/models/transformer_models.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2020, Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""The Transformer model for research baselines."""
+
+import tensorflow as tf
+import numpy as np
+from utils.models.stackoverflow_models import TransposableEmbedding
+
+
+def scaled_dot_product_attention(query: tf.Tensor, key: tf.Tensor, value: tf.Tensor, mask: tf.Tensor):
+  """Calculate the attention weights.
+
+  q (query), k (key), v (value) must have matching leading dimensions.
+  k, v must have matching penultimate dimension, i.e.: seq_len_k = seq_len_v.
+  The mask has different shapes depending on its type(padding or look ahead)
+  but it must be broadcastable for addition.
+
+  Args:
+    query: Query feature vectors, shape == (..., seq_len_q, depth).
+    key: Key feature vectors, shape == (..., seq_len_k, depth).
+    value: Value feature vectors, shape == (..., seq_len_v, depth_v).
+    mask: Float tensor with shape broadcastable
+          to (..., seq_len_q, seq_len_k). Defaults to None.
+  Returns:
+    output: The output attention vectors.
+    attention_weights: The attention weights.
+  """
+
+  matmul_qk = tf.matmul(query, key, transpose_b=True)
+
+  # scale matmul_qk
```

Comment:
We could remove this comment, or add more on why we need scaling. https://google.github.io/styleguide/pyguide.html#385-block-and-inline-comments